### PR TITLE
Support for MySQL and SSL Connections

### LINKED ISSUE #7039: Support for MySQL and SSL Connections
Feature Request

I'm not seeing any options in Piwik but enabling the `MySQL` connectors to use SSL. It's very important that we are given the capability of encrypting our connections to database servers, since they may not always be on the same server as Piwik.

### PR DESCRIPTION
**Pull request by ghost**
Hello,
My first ever open source pull request I am beginner but I use Piwik in work and want to make it even better, so please be nice but say if I did it wrong and how I should fix it.

I added the support of ssl for mysqli Adapter. From my point the needed options are ssl-ca and ssl-capath but this can vary on different platforms.

It would be nice to note that mysqli will silently fall back to non ssl connection if server don't support ssl.

Pls. let me know if this will be good, if yes i can think about adding options for sql settings to installer. 

MYSQLI_CLIENT_SSL in mysqli_real_connect is completely optional but would be nice to have it in case there will be some changes in mysqli.

fixes #7039

### CODE REVIEW

File: libs/Zend/Db/Adapter/Mysqli.php [hunk 1]
```diff
@@ -312,16 +312,64 @@ protected function _connect()
             }
         }
 
+        if($this->_config['usessl']) {
+            if(empty($this->_config['sslkey'])) {
+                $sslkey = null;
+            } else {
+                $sslkey = $this->_config['sslkey'];
+            }
```
Comment (sgiehl) [inline libs/Zend/Db/Adapter/Mysqli.php]: I would reformate this and the following if/else statements as it's shorter and easier to read:
```php
$sslkey = null;
if (!empty($this->_config['sslkey'])) {
    $sslkey = $this->_config['sslkey'];
}
```

File: libs/Zend/Db/Adapter/Mysqli.php [hunk 2]
```diff
@@ -312,16 +312,64 @@ protected function _connect()
...
+        } else {
+            $_isConnected = @mysqli_real_connect(
+                $this->_connection,
+                $this->_config['host'],
+                $this->_config['username'],
+                $this->_config['password'],
+                $this->_config['dbname'],
+                $port
+            );
+        }
```
Comment (sgiehl) [inline libs/Zend/Db/Adapter/Mysqli.php]: Maybe it would be simpler to use a ternary here, to avoid the code duplication:
```php
$_isConnected = @mysqli_real_connect(
   $this->_connection,
   $this->_config['host'],
   $this->_config['username'],
   $this->_config['password'],
   $this->_config['dbname'],
   $port,
   $this->config['usessl'] ? MYSQLI_CLIENT_SSL : null
);
```

**Comment by mattab** [comment]
Hi  @ktroska and thanks for this nice Pull request! 

Feedback:
 * can you add comments in the global.ini.php file to document "example values" for each of the `ssl*` settings? It's nice to show user what values are expected for each setting.
 - it would be nice to add SSL support in PDO Mysql as well as Mysqli (as we officially support both Mysqli and PDO)
 - maybe we can add a unit or integration test for this? not sure if it's easily do-able.
  * With such a test the advantage is that we would detect whenever the SSL support stops working in the future. Also if we have such a test, it will automatically run for both Mysqli and PDO as our tests in [CI server](https://travis-ci.org/piwik/piwik) run in both configurations.
 * to let the community know about this feature and how to use it, we will need to write a little FAQ "How do I configure Piwik to connect to Mysql database over secure SSL connection?"

**Comment by ktroska** [comment]
Hi @mattab Thx for feedback.
> can you add comments in the global.ini.php

Something like this:
>; Turn on or off SSL connection possible values for use_ssl: true or false
use_ssl = false
; Direct path to CA cert file, CA bundle supported (required for ssl connection)
ssl_ca =
; Direct path to client cert file (optional)
ssl_cert =
; Direct path to client key file (optional)
ssl_key =
; Direct path to CA cert files directory (optional)
ssl_ca_path =
; List of one or more ciphers for SSL encryption, in OpenSSL format (optional)
ssl_cipher =

I think something like that? Notice I think of adding underscores for all options for better visibility in both options and code.
Note: Requires PHP 5.3.7 http://php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants

>support in PDO Mysql 

code is done and tested I will merge it when we decide about global.ini.php content and underscores

> unit or integration test 

really first time I see this I will take a look and let you know if I can manage this, but since I started I want to make it to the end :)

> FAQ

I am thinking of integrating this into installer like SSL on and upload a ssl cert, the problem is there is no (i can think about) safe way to upload and store client key file if it is required, Or just FAQ info like this:

> To Enable secure connection with MySQL over ssl connection please add to file config/config.ini.php under [database] section:
<add here content of global.ini.php with explain as higher>
NOTE: Please not that if server is not supporting SSL, this function will **silently** fails to plain text connection. With this in mind this option can save you from sniffing packets, but **can't** be used as MITM protection (link to wikipedia? http://en.wikipedia.org/wiki/Man-in-the-middle_attack ).

Also we can do a test to check ssl support in connection like: 
SHOW STATUS LIKE 'Ssl_cipher';
!empty( Value )
Its still not protecting from mitm so I think its no use.

**Comment by mattab** [comment]
Another feedback: we should report to the user whether the secure connection works or not:

  * when SSL connection to DB works, in the system check ([screenshot](https://raw.githubusercontent.com/piwik/piwik-ui-tests/master/Installation_system_check.png)) we can display "Secure connection to database" with a green tick.   
  * when  a user has setup the SSL config settings, but somehow the SSL connection does not work (ie. one of the settings is wrong or the server SSL config is wrong), then ideally we want to detect and let user know. In the system check we would write "Warning: Secure connection could not be established!" next to "Secure connection to database".
  * if a user has not specified any the `ssl_*` INI settings, then user does not care about secure connection and the system check should not display the row "Secure connection to database".

**Comment by mattab** [comment]
> I think something like that?

it is easier to read :+1: 

> code is done and tested I will merge it when we decide about global.ini.php content and underscores

sounds good

> really first time I see this I will take a look and let you know if I can manage this, but since I started I want to make it to the end :)

Great. Automated testing is a powerful amazing concept for software development :-) Some pointers
 * Developer guide at: https://developer.piwik.org/guides/tests-php
 * Existing integration test you may be able to modify to add to: https://github.com/piwik/piwik/blob/master/tests/PHPUnit/Integration/DbTest.php

>      I am thinking of integrating this into installer like SSL on and upload a ssl cert

thanks for suggestion but it's not needed. FAQ will be good enough to document this feature. 
We could also add a link to the FAQ from: http://piwik.org/docs/how-to-secure-piwik/

> Also we can do a test to check ssl support in connection like:
> SHOW STATUS LIKE 'Ssl_cipher';
> !empty( Value )
> Its still not protecting from mitm so I think its no use.

It's still useful if Piwik lets user know whether the connection is done over SSL. See my previous comment about System check.


Finally:
 * in your PR you modify the libs/ Zend_Db files. We don't usually modify core lib files. I'm not sure how we should handle this situation. Maybe latest version of Zend_Db has support for SSL directly? Or do we maybe really need to modify the Zend_Db code itself?

**Comment by ktroska** [comment]
> Finally:

Well since I did mysqli first i did not notice that  Pdo can be easily done in piwik/core/Db/Adapter/Pdo/Mysql.php. So no Zend modification for this.

If it comes to Mysqli.... we need to call mysqli_ssl_set() between mysqli_init() and @mysqli_real_connect() 
And in zend there is only mysqli_options() which won't help us:
https://github.com/piwik/piwik/blob/master/libs/Zend/Db/Adapter/Mysqli.php#L300
same upstream:
https://github.com/zendframework/zend-db/blob/master/src/Adapter/Driver/Mysqli/Connection.php#L115
Maybe there is something I missed.
If there is no way I will think of changing the code to be compatible with Zend style and more object oriented and push it upstream to Zend meanwhile we will have it working in Piwik.

> system check 

I started working on it based on LoadDataInfileCheck.php after work and some sleep I think I will have it ready.

> Test 

I got general idea now to coding.

I can't thank you enough for sticking with me. :+1:

**Comment by ktroska** [comment]
I am now working on the automated testing, I will try to make it tomorrow (European time) but i am not sure if I will make it.
I think that all other things are done and ready.
Maybe I missed something, but I have tested it in my home lab and it looks ok. please confirm.

File: libs/Zend/Db/Adapter/Mysqli.php [hunk 3]
```diff
@@ -299,6 +299,15 @@ protected function _connect()
 
         $this->_connection = mysqli_init();
 
+        $use_ssl = false;
+        $ssl_options = array (
+            1012 => null, // 1012 = PDO::MYSQL_ATTR_SSL_CA
+            1013 => null, // 1013 = PDO::MYSQL_ATTR_SSL_CAPATH
+            1011 => null, // 1011 = PDO::MYSQL_ATTR_SSL_CERT
+            1014 => null, // 1014 = PDO::MYSQL_ATTR_SSL_CIPHER
+            1010 => null, // 1015 = PDO::MYSQL_ATTR_SSL_KEY
```
Comment (mattab) [inline libs/Zend/Db/Adapter/Mysqli.php]: * the comment says `1015` but code says `1010` 
 * maybe we could use these const directly instead of numeric values (in case they later change) eg. `PDO::MYSQL_ATTR_SSL_CIPHER` (if they are defined/available)

File: plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+namespace Piwik\Plugins\Diagnostics\Diagnostic;
+
+use Piwik\Common;
+use Piwik\Config;
+use Piwik\Db;
+use Piwik\Translation\Translator;
+
+/**
+ * Check if Piwik is connected with database through ssl.
+ * TODO: link to piwik FAQ into comment
```
Comment (mattab) [inline plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php]: > TODO: link to piwik FAQ into comment 

Sounds good!

File: plugins/Installation/lang/en.json [hunk 5]
```diff
@@ -67,6 +67,12 @@
         "SystemCheckCreateFunctionHelp": "Piwik uses anonymous functions for callbacks.",
         "SystemCheckDatabaseExtensions": "MySQL extensions",
         "SystemCheckDatabaseHelp": "Piwik requires either the mysqli extension or both the PDO and pdo_mysql extensions.",
+        "SystemCheckDatabaseSSL": "Database SSL Connection",
+        "SystemCheckDatabaseSSLCipher": "Used cipher",
+        "SystemCheckDatabaseSSLDisabled": "SSL support in your database server is disabled",
+        "SystemCheckDatabaseSSLNo": "Database server is not compiled with SSL support",
+        "SystemCheckDatabaseSSLNotWorking": "%s is set to %s but SSL connection is not working<br \/>",
```
Comment (mattab) [inline plugins/Installation/lang/en.json]: ideally translation strings don't include HTML markup -> remove `<br \/>`

File: plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    public function execute()
+    {
+        $use_ssl = Config::getInstance()->database['use_ssl'];
+        if (!$use_ssl) {
+            return array();
+        }
+
+        $label = $this->translator->translate('Installation_SystemCheckDatabaseSSL');
+
+        $cipher = Db::fetchAll("show status like 'Ssl_cipher'");
```
Comment (mattab) [inline plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php]: small detail: you can use `fetchOne` and then you don't need to do `$cipher[0]`

File: plugins/Installation/lang/en.json [hunk 7]
```diff
@@ -67,6 +67,12 @@
         "SystemCheckCreateFunctionHelp": "Piwik uses anonymous functions for callbacks.",
         "SystemCheckDatabaseExtensions": "MySQL extensions",
         "SystemCheckDatabaseHelp": "Piwik requires either the mysqli extension or both the PDO and pdo_mysql extensions.",
+        "SystemCheckDatabaseSSL": "Database SSL Connection",
+        "SystemCheckDatabaseSSLCipher": "Used cipher",
```
Comment (mattab) [inline plugins/Installation/lang/en.json]: suggest -> SSL Cipher being used

File: config/global.ini.php [hunk 8]
```diff
@@ -22,6 +22,26 @@
 type = InnoDB
 schema = Mysql
 
+; Database SSL Options START
+; Turn on or off SSL connection to database, possible values for use_ssl: true or false
+use_ssl = false
```
Comment (mattab) [inline config/global.ini.php]: Suggest to rename to `enable_ssl` as we tend to use `enable_*` more than `use_*`

File: plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        if(!empty($cipher[0]['Value'])) {
+             return array(DiagnosticResult::singleResult($label, DiagnosticResult::STATUS_OK, $this->translator->translate('Installation_SystemCheckDatabaseSSLCipher') . ': ' . $cipher[0]['Value']));
+        }
+
+        //no cipher, not working
+        $comment = sprintf($this->translator->translate('Installation_SystemCheckDatabaseSSLNotWorking'), "use_ssl", "true");
```
Comment (mattab) [inline plugins/Diagnostics/Diagnostic/DbOverSSLCheck.php]: in config file, we have habit of asking user to set 1 or 0 rather than true or false. Replace `true` with `1`

File: plugins/Installation/lang/en.json [hunk 10]
```diff
@@ -67,6 +67,12 @@
...
+        "SystemCheckDatabaseSSLDisabled": "SSL support in your database server is disabled",
+        "SystemCheckDatabaseSSLNo": "Database server is not compiled with SSL support",
+        "SystemCheckDatabaseSSLNotWorking": "%s is set to %s but SSL connection is not working<br \/>",
+        "SystemCheckDatabaseSSLOn": "Database server is supporting SSL connections, please check your configuration in config/config.ini.php <br /> Consider using REQUIRE SSL for your piwik database user",
```
Comment (mattab) [inline plugins/Installation/lang/en.json]: Suggested: `Your database supports SSL connections but SSL connection is not used. Check your database SSL settings in your Piwik config file.`

**Comment by mattab** [comment]
@ktroska it looks good. I left more comments. once you manage to add the automated test, it should be ready (or almost)!

**Comment by ktroska** [comment]
I have done test but i am not sure if it is ok, it was sure fun to learn something new and usefull.

Test:
Main problem is that we need to 'use PDO' which fails if it not exists and this way the whole test will fail badly, but if PDO is used all other things are tested ( like if used constants exists and if they values fits the ones used in code).
Also mysql test if there is a way to turn on ssl in mysql, since if the value would be NO(any other will be unexpected and test will fail) then you need to recompile mysql, test will fail but piwik will work just there will be no(?) way to connect to mysql through SSL.

File: core/Db/Adapter/Pdo/Mysql.php [hunk 11]
```diff
@@ -34,6 +34,23 @@ public function __construct($config)
         if (defined('PDO::MYSQL_ATTR_LOCAL_INFILE')) {
             $config['driver_options'][PDO::MYSQL_ATTR_LOCAL_INFILE] = true;
         }
+        if ($config['enable_ssl']) {
```
Comment (mattab) [inline core/Db/Adapter/Pdo/Mysql.php]: Is this change needed now that you've modified `core/Db/Adapter/Mysqli.php` maybe these file changes could be removed?

**Comment by mattab** [comment]
@ktroska as you can see in the Travis CI output for the job, the tests are failing: https://travis-ci.org/piwik/piwik/jobs/66525061

Here is how the test should work:
- the goal is to test that when the `ssl_*` config are set, then the Mysql connection over SSL works
- to test for this you need, in your test, to set those config flags before connecting to the database with `Db::createDatabaseObject` for example:

```
        $dbConfig = Config::getInstance()->database;
        $dbConfig['ssl_xxxx'] = xyz;
        Db::createDatabaseObject($dbConfig);

```

- after connection is successful, you can test that the mysql uses SSL, ideally by re-using the logic in `DbOverSSLCheck` diagnostic and asserting that SSL connection should be used.

Result:
- if connection to DB fails, test will fail
- if connection to DB is not done over SSL, test will fail
- if anything else errors, test will fail
- if DB connection works and SSL is used, then test will pass

As mentioned our CI runs on both PDO and Mysqli so this test(s) will automatically run on both.

**Comment by ktroska** [comment]
I have learned one thing from earlier testes, that constants in PDO have different values depending on build we have. So I changed the code to only use PDO constants or set values (those 1010-1014) only when using mysqli in case we don't have PDO support.

Those values work on my system and they can fail on any others so we can make them very high or change them to something different.

Since we are using SSL based constants from PDO this will require PHP 5.3.7 ( http://php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants ) this can't(?) be worked around when using PDO driver.

I wanted to do mysqli call easier and PDO compatible that's why I am using PDO constants values for mysqli, they can be changed to any other but it needs to be done in 3 places if you think its a good way changing it to some specially set values not PDO compatible than I can change it quickly.
* core/Db/Adapter/Mysqli.php 1x
* libs/Zend/Db/Adapter/Mysqli.php 2x

File: tests/PHPUnit/Integration/DbSSLTest.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+class DbSSLTest extends IntegrationTestCase
+{
+    public function testMysqlSSLConnection() {
+        $dbConfig = Config::getInstance()->database;
+        if(isset($dbConfig['enable_ssl']) AND $dbConfig['enable_ssl'] == true) {
+            Db::createDatabaseObject($dbConfig);
+            $cipher = Db::fetchRow("show status like 'Ssl_cipher'");
+            $this->assertNotEmpty($cipher['Value']);
+        } else {
+            $this->isTrue(true);
```
Comment (ktroska) [inline tests/PHPUnit/Integration/DbSSLTest.php]: either mark it true or mark it skipped with message like "enable_ssl = 0, test skipped"

**Comment by ktroska** [comment]
After some thinking:
* I will change the mysqli settings to not depend on PDO constants - use [driver_option][ssl_key] etc.
* change the enable_ssl to 1 in config to test the ssl calls in travis.
* remove require PHP version from tests.
    * split into one for mysqli, and other for mysql (add test for defined const - not for values since i know they are different and i dont need them)

**Comment by mattab** [comment]
Hi @ktroska 

We are still interested to have Mysql and SSL support in Piwik and this PR is very promising. What work is left to do in your opinion? if you need some help feel free to ask. Looking forward to having this one merged sometime :-)

**Comment by mattab** [comment]
@ktroska it would be possible to merge this Pull request in Piwik 3.0.0 - let us know if you are still interested :+1:

**Comment by mattab** [comment]
Closing the pull request as the work has stalled - it would be awesome if someone would finish this nice work!

**Comment by RafalLukawiecki** [comment]
Apologies for asking a question on a closed PR, but could @mattab (or someone else) clarify if Piwik 3.2.0 supports secure connections to its MySQL database? If so, what are the SSL (or otherwise) configuration parameters, please?

**Comment by sgiehl** [comment]
@RafalLukawiecki  no it's not implemented yet. See https://github.com/piwik/piwik/issues/7039

**Comment by mattab** [comment]
@RafalLukawiecki what we need is support from a developer in the community to finish this PR. we can re-open the PR if anyone will continue work on it :+1:
